lighthouse-cli/bin.ts

```diff
@@ -253,29 +252,47 @@ function handleError(err: LighthouseError) {
 }
 
 function runLighthouse(url: string,
-                       flags: {port: number, skipAutolaunch: boolean, selectChrome: boolean, output: any, outputPath: string, view: boolean},
+                       flags: {port: number, skipAutolaunch: boolean, selectChrome: boolean, output: any,
+                         outputPath: string, view: boolean, saveArtifacts: boolean, saveAssets: boolean},
                        config: Object): Promise<undefined> {
 
   let chromeLauncher: ChromeLauncher;
+  let performanceXServerPort: number;
   return initPort(flags)
+    .then(() => performanceXServer.startServer())
+    .then((port: number) => performanceXServerPort=port)
     .then(() => getDebuggableChrome(flags))
     .then(chrome => chromeLauncher = chrome)
     .then(() => lighthouse(url, flags, config))
+    .then((results: Results) => {
+      if (flags.saveArtifacts) {
+        assetSaver.saveArtifacts(results.artifacts);
+      }
+      if (flags.saveAssets) {
+        return assetSaver.saveAssets({url: results.url}, results.artifacts).then(() => results);
+      }
+      return results;
+    })
+    .then((results: Results) => {
+      // remove artifacts so users won't get unnecessary artifacts in the report.
+      results.artifacts = undefined;
+      return results;
+     })
     .then((results: Results) => Printer.write(results, flags.output, flags.outputPath))
     .then((results: Results) => {
-      const filename = `${assetSaver.getFilenamePrefix({url})}.report.html`;
-
       if (flags.output === Printer.OutputMode[Printer.OutputMode.pretty]) {
-        Printer.write(results, 'html', filename);
+        const filename = `${assetSaver.getFilenamePrefix({url})}.report.html`;
+        return Printer.write(results, 'html', filename);
       }
```

Comment:
Fixed a concurrency error here. Previously, promise will resolve to next 'then' before Printer finishes writing. No sure if that matters.